### PR TITLE
Add Image To Text Generation pipeline

### PR DESCRIPTION
**Pull request by OlivierDehaene**
# What does this PR do?

Add Image To Text Generation pipeline. The pipeline currently defaults to [nlpconnect/vit-gpt2-image-captioning](https://huggingface.co/nlpconnect/vit-gpt2-image-captioning).

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#start-contributing-pull-requests),
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [x] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [x] Did you write any new necessary tests?

## Who can review?

This features was asked for by @Narsil.

### CODE REVIEW

File: src/transformers/models/auto/feature_extraction_auto.py
Comment (OlivierDehaene) [inline src/transformers/models/auto/feature_extraction_auto.py]: @Narsil,

I am not sure about this. There is nothing that "forces" future vision-encoder-decoder to use VIT as an encoder. But if we remove it, we loses support for https://huggingface.co/ydshieh/vit-gpt2-coco-en-ckpts

File: src/transformers/pipelines/__init__.py
Comment (OlivierDehaene) [inline src/transformers/pipelines/__init__.py:329]: @Narsil, 

Is this correct way of adding a model which can have multiple tokenizers?

For example, [this model](https://huggingface.co/nlpconnect/vit-gpt2-image-captioning) uses `GPT2Tokenizer` as a tokenizer whereas [this one](https://huggingface.co/microsoft/trocr-base-printed) uses `RobertaTokenizer` but both are vision encoder decoders.

File: tests/pipelines/test_pipelines_image2text_generation.py
Comment (OlivierDehaene) [inline tests/pipelines/test_pipelines_image2text_generation.py:71]: This must be added before merging.

File: src/transformers/pipelines/__init__.py
Comment (Narsil) [reply src/transformers/pipelines/__init__.py:335]: `AutoTokenizer` should be able to handle it by looking at the config.

There is afaik only 1 type of model which actually uses really two different tokenizers for the same model, which are the Marian variants. It's dealt by swapping the underlying tokenizer when necessary (not great, but it does the trick).

`MULTI_MODEL_CONFIGS` are extremely general since the idea is just to use in coordination two different kind of models, that's why the pipeline needs to know if we're looking at such models, because for such models, we need to try and load `tokenizer` (which should work out of the box) without really looking too closely at the arch. (We could do that everywhere, but this auto loading was written this way because some models were misconfigured and this was the only way to properly load the pipeline despite misconfigurations).

File: src/transformers/pipelines/image2text_generation.py
Comment (Narsil) [inline src/transformers/pipelines/image2text_generation.py:73]: I think these come from `text-generation` pipeline.

I am not sure we should enable them before it's necessary, returning tensors is not really aligned with the goal to focus on non ML practictionners.

I think returning pure text is better as a first pass (if users ask for more parameters, it's always easier to add later than be encumbered by the weight of being forced to support some parameters, and most importantly the parameter matrix that arises fast).

File: src/transformers/pipelines/image2text_generation.py
Comment (Narsil) [inline src/transformers/pipelines/image2text_generation.py:90]: > , which must then be passed as a string.

Is that true ? 

> Images in a batch must all be in the same format: all as http links, all as local paths, or all as PIL
                images.

I think even that isn't true anymore, which pipeline did you take that from ? We might want to update since I think there's no such limit anymore.

File: src/transformers/pipelines/image2text_generation.py
Comment (Narsil) [inline src/transformers/pipelines/image2text_generation.py:96]: I think we can remove the return_tensors here along with the parameter.

File: tests/pipelines/test_pipelines_image2text_generation.py
Comment (Narsil) [inline tests/pipelines/test_pipelines_image2text_generation.py:60]: This is probably bugged, since the pipelines don't work with `flax`.

I am guessing this ends up using `pytorch` under the hood on your local computer.

you can force a framework by doing `pipeline("image2text-generation", model="hf-internal-testing/tiny-random-vit-gpt2", framework="flax")`

This should fail.

And I don't think the model tester even attempts to run the pipeline tests with flax on.

File: tests/pipelines/test_pipelines_image2text_generation.py
Comment (Narsil) [inline tests/pipelines/test_pipelines_image2text_generation.py:67]: Despite being random when initialized, this model is actually deterministic (or should be).
So we need to test against the actual string, not `ANY(str)`.

For the general test, we are testing with the same code multiple different models, so we cannot know which string we are testing against and are limited to testing against output shape.

For these specific tests, we can and we should be more specific.

File: tests/pipelines/test_pipelines_image2text_generation.py
Comment (Narsil) [inline tests/pipelines/test_pipelines_image2text_generation.py:91]: > 
> 
> Despite being random when initialized, this model is actually deterministic (or should be).
> So we need to test against the actual string, not ANY(str).
> 
> For the general test, we are testing with the same code multiple different models, so we cannot know which string we are testing against and are limited to testing against output shape.
> 
> For these specific tests, we can and we should be more specific.

Same here

File: tests/pipelines/test_pipelines_image2text_generation.py
Comment (Narsil) [inline tests/pipelines/test_pipelines_image2text_generation.py:118]:
```suggestion
            outputs, [[{"generated_text": "a cat laying on a blanket next to a cat laying on a bed "}], [{"generated_text": "a cat laying on a blanket next to a cat laying on a bed "}]]
```
I tend to prefer overly specific in tests as it reduces strain on the reader to know if the test is ok or not.
Also it would be better if we had 2 different images to check against ,but we don't have in fixtures (we're mindful of the repo size) so this is OK like this.

File: src/transformers/pipelines/__init__.py
Comment (Narsil) [inline src/transformers/pipelines/__init__.py:315]: Why are the two models different ?
 
 Which one is better @ydshieh do you have an advice here maybe ?
 
 Whenever possible we put the best default model here, usually it's the same for both TF and PT, but it depends of implementations.

**Review by Narsil — COMMENTED** [review]
Very nice first PR.

It looks very nice already.

@mishig25 I don't think we have a widget for text-to-image, do we ? Do you think we could add one ?

@sgugger for a second opinion on this pipeline.

File: src/transformers/pipelines/__init__.py
Comment (ydshieh) [reply src/transformers/pipelines/__init__.py:315]: `"nlpconnect/vit-gpt2-image-captioning"` mentions 

> This is an image captioning model training by @ydshieh in flax, this is pytorch version of https://huggingface.co/ydshieh/vit-gpt2-coco-en-ckpts model.

So I believe it's just the same checkpoint. In `ydshieh/vit-gpt2-coco-en` there is also PyTorch checkpoint too, so we can use the same repo.

Comment (OlivierDehaene) [reply src/transformers/pipelines/__init__.py:315]: Should I just use `ydshieh/vit-gpt2-coco-en` as a default for both PyTorch and Tensorflow then? Or should we add the tf weights to `nlpconnect/vit-gpt2-image-captioning` which seems more popular.

File: src/transformers/pipelines/image2text_generation.py
Comment (OlivierDehaene) [reply src/transformers/pipelines/image2text_generation.py]: Ok I simplified the generation.

File: tests/pipelines/test_pipelines_image2text_generation.py
Comment (OlivierDehaene) [reply tests/pipelines/test_pipelines_image2text_generation.py]: I removed flax support from the pipeline.

File: src/transformers/pipelines/image2text_generation.py
Comment (OlivierDehaene) [reply src/transformers/pipelines/image2text_generation.py]: I removed this warning from the docstring. Indeed, it was not necessary anymore.

File: src/transformers/pipelines/__init__.py
Comment (OlivierDehaene) [reply src/transformers/pipelines/__init__.py:315]: @ydshieh,
It seems that the `preprocessor_config.json` in `ydshieh/vit-gpt2-coco-en` is missing an entry. I opened a PR [here](https://huggingface.co/ydshieh/vit-gpt2-coco-en/discussions/1) to fix it.

File: src/transformers/pipelines/image2text_generation.py
Comment (OlivierDehaene) [inline src/transformers/pipelines/image2text_generation.py:82]: @Narsil,
What do you think of this? The issue is that `model_inputs` is a dict with only a `pixel_values` entry. While PyTorch generate utils is able to understand that `pixel_values` must be matched with the encoder input, Tensorflow's isn't.

File: tests/pipelines/test_pipelines_image2text_generation.py
Comment (OlivierDehaene) [inline tests/pipelines/test_pipelines_image2text_generation.py:139]: This fails at the moment because [this PR](https://huggingface.co/ydshieh/vit-gpt2-coco-en/discussions/1) must be merged first.

Comment (ydshieh) [reply tests/pipelines/test_pipelines_image2text_generation.py:139]: merged that PR! Thanks.

**Review by sgugger — APPROVED** [review]
Thanks for adding this new pipeline!

File: src/transformers/pipelines/image2text_generation.py
Comment (Narsil) [reply src/transformers/pipelines/image2text_generation.py:82]: It's a nice workaround.

It would be better if it wasn't there, I would ask @gante about this (Worked on TF generate quite extensively).
I expect it's a TF limit that it needs to be consistent in the naming hence the issue about `input_ids`.

But maybe there's cleaner code overall to be done in the lib.

**Review by Narsil — APPROVED** [review]
LGTM

**Comment by NielsRogge** [comment]
cc @mishig25, the inference widgets for models like TrOCR, Donut can now be created! 🥳

File: src/transformers/pipelines/image2text_generation.py
Comment (gante) [reply src/transformers/pipelines/image2text_generation.py:82]: Hi @OlivierDehaene @Narsil 👋 This seems to be a limitation on the TF generation side. I've taken note of the issue, including the removal of this line :D

**Pull request merged**
